pkg/compact/compact_e2e_test.go

```diff
@@ -393,6 +387,122 @@ func TestGroup_Compact_e2e(t *testing.T) {
 	})
 }
 
+func TestGroup_GroupConcurrency_e2e(t *testing.T) {
+	objtesting.ForeachStore(t, func(t *testing.T, bkt objstore.Bucket) {
+		ctx, cancel := context.WithTimeout(context.Background(), 120*time.Second)
+		defer cancel()
+
+		// Create fresh, empty directory for actual test.
+		dir, err := ioutil.TempDir("", "test-compact")
+		testutil.Ok(t, err)
+		defer func() { testutil.Ok(t, os.RemoveAll(dir)) }()
+
+		logger := log.NewLogfmtLogger(os.Stderr)
+
+		reg := prometheus.NewRegistry()
+
+		ignoreDeletionMarkFilter := block.NewIgnoreDeletionMarkFilter(logger, objstore.WithNoopInstr(bkt), 48*time.Hour, fetcherConcurrency)
+		duplicateBlocksFilter := block.NewDeduplicateFilter()
+		metaFetcher, err := block.NewMetaFetcher(nil, 32, objstore.WithNoopInstr(bkt), "", nil, []block.MetadataFilter{
+			ignoreDeletionMarkFilter,
+			duplicateBlocksFilter,
+		}, nil)
+		testutil.Ok(t, err)
+
+		blocksMarkedForDeletion := promauto.With(nil).NewCounter(prometheus.CounterOpts{})
+		garbageCollectedBlocks := promauto.With(nil).NewCounter(prometheus.CounterOpts{})
+		sy, err := NewSyncer(nil, nil, bkt, metaFetcher, duplicateBlocksFilter, ignoreDeletionMarkFilter, blocksMarkedForDeletion, garbageCollectedBlocks, 5)
+		testutil.Ok(t, err)
+
+		comp, err := tsdb.NewLeveledCompactor(ctx, reg, logger, []int64{1000, 3000}, nil)
+		testutil.Ok(t, err)
+
+		planner := NewTSDBBasedPlanner(logger, []int64{1000, 3000})
+
+		grouper := NewDefaultGrouper(logger, bkt, false, true, reg, blocksMarkedForDeletion, garbageCollectedBlocks, planner)
+		bComp, err := NewBucketCompactor(logger, sy, grouper, planner, comp, dir, bkt, 2)
+		testutil.Ok(t, err)
+		extLabels := labels.Labels{{Name: "e1", Value: "1"}}
+		metas := createAndUpload(t, bkt, []blockgenSpec{
```

Comment:
This is great, but I would also vote for smaller grouper/planner unit test, it's quite critical path (compactions are irreversible)

Reply:
agreed